File: Students/controller.py
from cmath import log
from os import access
from flask import request, jsonify
from Utilities.db_operations import *
from Utilities.error_handler import custom_response_maker
from .validation import *
import hashlib
from flask_jwt_extended import create_access_token, decode_token


def create_user():
    body = {}
    body = request.get_json()

    if(isinstance(validate_insert(body), dict)):
        return custom_response_maker(validate_insert(body))

    plaintext = hashlib.md5(body['password'].encode())
    hashed_password = plaintext.hexdigest()
    try:
        query = 'INSERT INTO Students VALUES (%d , "%s" , "%s" , "%s" );' % (
                body['id'], body['username'], body['email'], hashed_password)
    except KeyError as e:
        remodeled_err = {
            'err_code': 'Key Error',
            'message': 'Key error requires %s key' % (e.args[0])
        }
        return custom_response_maker(remodeled_err)
    return custom_response_maker(insert_row(query))


def show_users():
    if authorize(request):
        try:
            query = 'SELECT * FROM Students'
        except KeyError as e:
            remodeled_err = {
                'err_code': 'Key Error',
                'message': 'Key error requires %s key' % (e.args[0])
            }
            return custom_response_maker(remodeled_err)
        return custom_response_maker(read_row(query))
    else:
        return { "msg" : "unauthorized user"}



def login_user():
    body = {}
    body = request.get_json()

    if(isinstance(validate_login(body), dict)):
        return custom_response_maker(validate_login(body))

    try:
        query = 'SELECT password FROM Students WHERE username="%s"' % (
            body['username'])
    except KeyError as e:
        remodeled_err = {
            'err_code': 'Key Error',
            'message': 'Key error requires %s key' % (e.args[0])
        }
        return custom_response_maker(remodeled_err)

    res = read_row(query)
    login_res = validate_password(res['data'][0]['password'],body)
    if login_res["is_success"]:
        access_token = create_access_token(identity=body['username'])
        return access_token
    else:
        return login_res


def authorize(req):
    try:
        token = req.headers['AUTHORIZATION'].split(' ')[1]
    except KeyError as e:
        print(req.headers)
        return False
    username = decode_token(token)['sub']

    query = 'SELECT username FROM Students WHERE username="%s";'%(username)
    data= read_row(query)
    db_user = data['data'][0]['username']

    if db_user == username:
        return True
    else:
        return False

File: Students/controller.js
const md5 = require("md5");
const db = require("../Utilities/db-operations");
const { customResponse } = require("../Utilities/custom-response");
const { createToken } = require("../Utilities/authorization");

const getInsert = async (req, res, next) => {
  try {
    const hash = md5(req.body.password);
    let body = [
      req.body.id,
      req.body.username,
      req.body.email,
      hash,
      req.body.phone_no,
    ];

    let query =
      "INSERT INTO Students(id,username,email,password,phone_no) VALUES (?)";
    let result = await db
      .executeQuery(query, [body])
      .catch(function reject(err) {
        next(Error(err.sqlMessage));
        return null;
      });

    if (result !== null) {
      let user = { username: req.body.username, id: req.body.id };
      let token = createToken(user);
      let newresult = {
        code: 200,
        message: `${result.affectedRows} student added to db`,
        token: token,
      };

      return customResponse(newresult, res);
    }
  } catch (err) {
    next(err);
  }
};

const getInsertMultiple = async (req, res, next) => {
  try {
    let users = [];
    req.body.forEach((user) => {
      let hash = md5(user.password);
      users.push([user.id, user.username, user.email, hash, user.phone_no]);
    });

    let query =
      "INSERT INTO Students(id,username,email,password,phone_no) VALUES ?";
    let result = await db
      .executeQuery(query, [users])
      .catch(function reject(err) {
        next(Error(err.sqlMessage));
        return null;
      });

    if (result !== null) return customResponse(result, res);
  } catch (err) {
    next(err);
  }
};

const getUpdate = async (req, res, next) => {
  try {
    let body = [req.body.new_value];
    let query = `UPDATE Students SET ${req.body.column} = ? WHERE id= ${req.userData.id} `;

    let result = await db.executeQuery(query, body).catch(function reject(err) {
      next(Error(err.sqlMessage));
      return null;
    });

    if (result !== null) return customResponse(result, res);
  } catch (err) {
    next(err);
  }
};

const getDelete = async (req, res, next) => {
  try {
    body = [req.userData.id];
    let query = "DELETE FROM Students WHERE id=?";

    let result = await db.executeQuery(query, body).catch(function reject(err) {
      next(Error(err.sqlMessage));
      return null;
    });

    if (result !== null) return customResponse(result, res);
  } catch (err) {
    next(err);
  }
};

const getSelect = async (req, res, next) => {
  try {
    let body = req.params._id;
    let query = "SELECT * FROM Students WHERE id=?";

    let result = await db.executeQuery(query, body).catch(function reject(err) {
      next(Error(err.sqlMessage));
      return null;
    });

    if (result !== null) {
      result.currentUser = req.userData.username;
      return customResponse(result, res);
    }
  } catch (err) {
    next(err);
  }
};

const getSelectAll = async (req, res, next) => {
  try {
    let body = [];
    let query = "SELECT * FROM Students";
    let result = await db.executeQuery(query, body).catch(function reject(err) {
      next(Error(err.sqlMessage));
      return null;
    });

    if (result !== null) {
      result.currentUser = req.userData.username;
      return customResponse(result, res);
    }
  } catch (err) {
    next(err);
  }
};

const getLogin = async (req, res, next) => {
  try {
    let username = req.body.username;
    let user_pass = req.body.password;

    let query = "SELECT * FROM Students WHERE username = ?";
    let body = [username];

    let result = await db.executeQuery(query, body).catch(function reject(err) {
      next(Error(err.sqlMessage));
      return null;
    });

    if (result !== null) {
      //check if user exists
      if (result.length === 0) {
        next(Error("user does not exist"));
      } else {
        let user = { username: req.body.username, id: result[0].id };
        let token = createToken(user);
        password = result[0].password;
        let hash = md5(user_pass);
        if (password === hash) {
          return customResponse(
            { code: 200, message: "login successful", token: token },
            res
          );
        } else {
          next(Error("login unsuccesfull"));
        }
      }
    }
  } catch (err) {
    next(err);
  }
};

module.exports = {
  getInsert,
  getUpdate,
  getDelete,
  getSelect,
  getSelectAll,
  customResponse,
  getLogin,
  getInsertMultiple,
};
